### PR TITLE
Full captions

### PR DESCRIPTION
**Pull request by alexandrewalsh**
Full caption transcript available upon pressing the CC button next to the bookmark button.
- appears as a scrollable underneath the bookmarks, size of which is determined by screen size (size = 25% of screen size)
- all text is concatenated
- i've noticed sometimes the captions that we pull are actually different from the captions displayed on the video? I think this is because the channel i have been using, Kurzgesagt, actually curates its own captions
EDIT: Realized the reason the captions are different is because the ones displayed are the English Caption Track, while the ones I was using are the Arabic Caption Track translated to English. Should be fixed in retrieve-captions PR.

### CODE REVIEW

File: src/main/webapp/js/captions.js [hunk 1]
```diff
@@ -56,11 +68,15 @@ function execute(url) {
...
     try {
         videoId = getIdFromUrl(url);
+        currentVideoID = videoId;
     } catch {
         renderError("Invalid youtube url!");
         return;
     }
 
+    // hides the bookmarks for the previous video
+    $('#bookmark-display-div').html('');
```
Comment (RickyAvina) [inline src/main/webapp/js/captions.js:78]: it may be better to use JQuery's `.hide()`

File: src/main/webapp/js/captions.js [hunk 2]
```diff
@@ -78,6 +94,7 @@ function execute(url) {
         return;
     }
 
+    video_id = videoId;
```
Comment (RickyAvina) [inline src/main/webapp/js/captions.js:97]: The naming may cause confusion

Comment (alexandrewalsh) [reply src/main/webapp/js/captions.js]: changed to current_vID

**Comment by jiashenwang** [comment]
Since the notes CL is merged. Could we rebase this PR?

File: src/main/webapp/js/captions.js [hunk 3]
```diff
@@ -541,6 +550,30 @@ function setBookmarkButton() {
...
+    document.getElementById("FullCap").innerHTML = "";
+
+    // Display captions on click
+    $('#fullcap-button').click(function() {
+        // checks to see if captions already exist in the database
+        if (document.getElementById("FullCap").innerHTML == "") {
+            fetch('/fullcaption?id=' + current_vID, {
+                method: 'GET',
+            }).then((response) => response.text()).then((text) => {
+                if (text.length > 0) {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:565]: ```if (text != null && text.trim() != '')```

File: src/main/webapp/js/captions.js [hunk 4]
```diff
@@ -541,6 +550,30 @@ function setBookmarkButton() {
...
+                    // Sets the results table
+                    document.getElementById("FullCap").innerHTML = text;
+                    console.log("Fetching full captions from database...");
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:568]: I believe the full caption has been fetched at this point right?

File: src/main/webapp/js/captions.js [hunk 5]
```diff
@@ -541,6 +550,30 @@ function setBookmarkButton() {
...
+    $("#timestamp-timeline").append('<button id="fullcap-button"><i style="font-size:24px" class="fa fa-cc"></i></button>');
+    document.getElementById("FullCap").innerHTML = "";
+
+    // Display captions on click
+    $('#fullcap-button').click(function() {
+        // checks to see if captions already exist in the database
+        if (document.getElementById("FullCap").innerHTML == "") {
+            fetch('/fullcaption?id=' + current_vID, {
+                method: 'GET',
+            }).then((response) => response.text()).then((text) => {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:564]: Let's also handle the error case.

File: src/main/java/com/google/sps/servlets/CaptionRetrievalServlet.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        CaptionStorageManager dbi = new CaptionStorageManager();
+        Gson gson = new Gson();
+        String videoID = (String) request.getParameter(REQUEST_ID_PARAM);
+        response.setContentType(RESPONSE_JSON_CONTENT);
+        String captionText = "";
+
+        try {
+            if (videoID != null && dbi.videoInDb(videoID)) {
```
[the remark this answers is not in the document]
Comment (alexandrewalsh) [reply src/main/java/com/google/sps/servlets/CaptionRetrievalServlet.java:60]: interesting, makes sense though. done

File: src/main/webapp/js/captions.js (hunk 3 above)
Comment (alexandrewalsh) [reply src/main/webapp/js/captions.js:565]: we do this to illustrate what exactly we're looking out for? and `text.length > 0` is too broad?

File: src/main/webapp/js/captions.js (hunk 4 above)
Comment (alexandrewalsh) [reply src/main/webapp/js/captions.js:568]: yeah, log was just for me i had just copied it from another log message, will take it out.

File: src/main/webapp/WEB-INF/appengine-web.xml [hunk 7]
```diff
@@ -2,6 +2,7 @@
 <appengine-web-app xmlns="http://appengine.google.com/ns/1.0">
   <threadsafe>false</threadsafe>
   <sessions-enabled>true</sessions-enabled>
+  <module>alwa</module>
```
Comment (RickyAvina) [inline src/main/webapp/WEB-INF/appengine-web.xml:5]: use default or don't include

Comment (alexandrewalsh) [reply src/main/webapp/WEB-INF/appengine-web.xml]: done

**Pull request merged**